### PR TITLE
Fixed error not being raised when an invalid parent is provided 

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #4010 
# What's Changed

1.  Raise an exception in `IPAddress.clean` if `parent` is specified but it's not the closest `prefix` to the `IPAddress`.
2. Changed `IPAddressSerializer.validate` to use "IPAddress.clean" to eliminate code duplication.

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Review by bryanculver — APPROVED** [review]
@timizuoebideri1 please review. It seems with the model save changes something broke in the other test cases.

**Review by bryanculver — DISMISSED** [review]
@timizuoebideri1 please review. It seems with the model save changes something broke in the other test cases.

File: nautobot/ipam/api/serializers.py [hunk 1]
```diff
@@ -265,6 +265,16 @@ def validate(self, data):
         if self.instance is None and not any([namespace, parent]):
             raise ValidationError({"__all__": "One of parent or namespace must be provided"})
 
+        # DRF does not call Model clean method by default; however, to avoid repeating validation checks in Model clean, call clean method here.
```
Comment (timizuoebideri1) [inline nautobot/ipam/api/serializers.py:268]:
```suggestion
        # DRF does not call Model clean method by default; however, to avoid rewriting validation checks in Model clean, call clean method here.
```

File: nautobot/ipam/api/serializers.py [hunk 2]
```diff
@@ -265,6 +265,16 @@ def validate(self, data):
...
+        instance = self.instance or self.Meta.model(**data)
+        try:
+            instance.clean()
+        except ValidationError as e:
+            raise serializers.ValidationError(e.args[0])
+
+        # `dns_name` and `parent` were been modified in the Model clean method, pass those modified data into `data`
```
Comment (timizuoebideri1) [inline nautobot/ipam/api/serializers.py:275]:
```suggestion
        # `dns_name` and `parent` has been modified in the Model clean method, pass those modified data into `data`
```

File: nautobot/ipam/models.py [hunk 3]
```diff
@@ -1079,30 +1097,26 @@ def clean(self):
         if self.type == choices.IPAddressTypeChoices.TYPE_SLAAC and self.ip_version != 6:
             raise ValidationError({"type": "Only IPv6 addresses can be assigned SLAAC type"})
 
+        # Validate `parent` can be used as a parent for this ipaddress
```
Comment (timizuoebideri1) [inline nautobot/ipam/models.py:1100]:
```suggestion
        # Validate `parent` can be used as the parent for this ipaddress
```

File: nautobot/ipam/models.py [hunk 4]
```diff
@@ -1079,30 +1097,26 @@ def clean(self):
...
-            .get_closest_parent(self.host, include_self=True)
-        )
-
         # 3.0 TODO: uncomment the below to enforce this constraint
         # if self.parent.type != choices.PrefixTypeChoices.TYPE_NETWORK:
         #     err_msg = f"IP addresses cannot be created in {self.parent.type} prefixes. You must create a network prefix first."
         #     raise ValidationError({"address": err_msg})
 
+        # validated_save is not always called especially in Test environment;
+        # but we still need to set the self.parent attr in this cases
```
Comment (timizuoebideri1) [inline nautobot/ipam/models.py:1117]:
```suggestion
        # but we still need to set the `self.parent` attr in these cases
```

File: nautobot/ipam/api/serializers.py [hunk 5]
```diff
@@ -265,6 +265,16 @@ def validate(self, data):
         if self.instance is None and not any([namespace, parent]):
             raise ValidationError({"__all__": "One of parent or namespace must be provided"})
 
+        # DRF does not call Model clean method by default; however, to avoid rewriting validation checks in Model clean, call clean method here.
```
Comment (glennmatthews) [inline nautobot/ipam/api/serializers.py:268]: DRF doesn't call model clean by default, but Nautobot's `ValidatedModelSerializer` (parent of `NautobotModelSerializer` and therefore of this serializer) does do so. Probably a more correct fix here would be to call `super.validate(data)` at some point, rather than calling `instance.clean()` directly?

File: nautobot/ipam/models.py [hunk 6]
```diff
@@ -1043,8 +1043,7 @@ class Meta:
     def __init__(self, *args, **kwargs):
         address = kwargs.pop("address", None)
         namespace = kwargs.pop("namespace", None)
-        # We don't want users providing their own parent since it will be derived automatically.
-        parent = kwargs.pop("parent", None)
+        parent = kwargs.get("parent", None)
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1046]: Why the change from `pop` to `get`? Does the comment about not allowing users to directly set `parent` no longer apply because of the `clean` changes?

File: nautobot/ipam/models.py [hunk 7]
```diff
@@ -1066,6 +1065,25 @@ def _deconstruct_address(self, address):
 
     natural_key_field_names = ["parent__namespace", "host"]
 
+    def _get_closest_parent(self):
+        if self.present_in_database:
+            namespace = self._namespace or self.parent.namespace
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1070]: I can't see a case in `__init__()` where parent is non-None but self._namespace is None - is the "or" actually needed here?

File: nautobot/ipam/models.py [hunk 8]
```diff
@@ -1066,6 +1065,25 @@ def _deconstruct_address(self, address):
...
+        elif self._namespace is None:
+            raise ValidationError({"parent": "Namespace could not be determined."})
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1072]: When would this error be triggered? I don't see it in the tests.

File: nautobot/ipam/models.py [hunk 9]
```diff
@@ -1066,6 +1065,25 @@ def _deconstruct_address(self, address):
...
+            namespace = self._namespace
+
+        try:
+            closes_parent = (
+                Prefix.objects.filter(namespace=namespace)
+                # 3.0 TODO: disallow IPAddress from parenting to a TYPE_POOL prefix, instead pick closest TYPE_NETWORK
+                # .exclude(type=choices.PrefixTypeChoices.TYPE_POOL)
+                .get_closest_parent(self.host, include_self=True)
+            )
+            return closes_parent
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1083]:
```suggestion
            closest_parent = (
                Prefix.objects.filter(namespace=namespace)
                # 3.0 TODO: disallow IPAddress from parenting to a TYPE_POOL prefix, instead pick closest TYPE_NETWORK
                # .exclude(type=choices.PrefixTypeChoices.TYPE_POOL)
                .get_closest_parent(self.host, include_self=True)
            )
            return closest_parent
```

File: nautobot/ipam/models.py [hunk 10]
```diff
@@ -1079,30 +1097,26 @@ def clean(self):
         if self.type == choices.IPAddressTypeChoices.TYPE_SLAAC and self.ip_version != 6:
             raise ValidationError({"type": "Only IPv6 addresses can be assigned SLAAC type"})
 
+        # Validate `parent` can be used as the parent for this ipaddress
+        if self.address and self.parent:
+            closes_parent = self._get_closest_parent()
+            if self.parent != closes_parent:
+                raise ValidationError({"parent": f"{self.parent} cannot be assigned as a parent."})
+            self.parent = closes_parent
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1105]:
```suggestion
            closest_parent = self._get_closest_parent()
            if self.parent != closest_parent:
                raise ValidationError({"parent": f"{self.parent} cannot be assigned as a parent."})
            self.parent = closest_parent
```

File: nautobot/ipam/models.py [hunk 11]
```diff
@@ -1079,30 +1097,26 @@ def clean(self):
         if self.type == choices.IPAddressTypeChoices.TYPE_SLAAC and self.ip_version != 6:
             raise ValidationError({"type": "Only IPv6 addresses can be assigned SLAAC type"})
 
+        # Validate `parent` can be used as the parent for this ipaddress
+        if self.address and self.parent:
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1101]: when would `self.address` evaluate as falsey?

File: nautobot/ipam/models.py [hunk 12]
```diff
@@ -1079,30 +1097,26 @@ def clean(self):
...
-
         # 3.0 TODO: uncomment the below to enforce this constraint
         # if self.parent.type != choices.PrefixTypeChoices.TYPE_NETWORK:
         #     err_msg = f"IP addresses cannot be created in {self.parent.type} prefixes. You must create a network prefix first."
         #     raise ValidationError({"address": err_msg})
 
+        # validated_save is not always called especially in Test environment;
+        # but we still need to set the `self.parent` attr in these cases
+        if self.address and not self.parent:
+            self.parent = self._get_closest_parent()
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1119]: Does this mean that I can now do something like `IPAddress.objects.create(address="10.1.1.1", parent=Prefix.objects.get(network="172.16.0.0/16"))` and as long as I never call clean() it'll be accepted by the DB? We definitely should still prevent that.

File: nautobot/ipam/models.py (hunk 6 above)
Comment (timizuoebideri1) [reply nautobot/ipam/models.py:1046]: Yes, because if it is popped, we will no longer be able to validate it in the clean method.

File: nautobot/ipam/models.py (hunk 8 above)
Comment (timizuoebideri1) [reply nautobot/ipam/models.py:1072]: if neither `namespace` nor `parent` is specified when an IPAddress is created e.g `IPAddress.objects.create(address="...", status="...")`

File: nautobot/ipam/models.py (hunk 12 above)
Comment (timizuoebideri1) [reply nautobot/ipam/models.py:1119]: This would means we would still have to silently discard invalid parent, which is what the issue is reporting against.
Except we force validation save for `IPAddress`

File: nautobot/ipam/forms.py [hunk 13]
```diff
@@ -493,6 +493,8 @@ def clean_namespace(self):
         setattr(self.instance, "_namespace", namespace)
 
     def clean(self):
+        # Pass address to the instance, because this is required to be accessible in the IPAddress.clean method
+        self.instance.address = self.cleaned_data.get("address")
```
Comment (glennmatthews) [inline nautobot/ipam/forms.py:497]: Is this specifically for "create", for "edit", or both cases that this is needed?

Comment (timizuoebideri1) [reply nautobot/ipam/forms.py:497]: Both cases, cause address might/can be updated

File: nautobot/ipam/models.py [hunk 14]
```diff
@@ -1079,30 +1089,31 @@ def clean(self):
...
-            Prefix.objects.filter(namespace=namespace)
-            # 3.0 TODO: disallow IPAddress from parenting to a TYPE_POOL prefix, instead pick closest TYPE_NETWORK
-            # .exclude(type=choices.PrefixTypeChoices.TYPE_POOL)
-            .get_closest_parent(self.host, include_self=True)
-        )
+        # Validate `parent` can be used as the parent for this ipaddress
+        if self.parent:
+            closest_parent = self._get_closest_parent()
+            if self.parent != closest_parent:
+                raise ValidationError({"parent": f"{self.parent} cannot be assigned as a parent."})
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1100]: Perhaps something more explicit? Like:

```suggestion
                raise ValidationError(
                    {
                        "parent": (
                            f"{self.parent} cannot be assigned as the parent of {self}. "
                            f" In namespace {self._namespace}, the expected parent would be {closest_parent}."
                        )
                    }
                )
```

File: nautobot/ipam/tests/test_models.py [hunk 15]
```diff
@@ -954,6 +957,28 @@ def test_parenting_constraints(self):
...
+        with self.assertRaises(ValidationError) as err:
+            ipaddress = IPAddress(address="192.168.0.1/16", parent=prefixes[0], status=self.status)
+            ipaddress.validated_save()
+        self.assertEqual(f"{prefixes[0]} cannot be assigned as a parent.", err.exception.message_dict["parent"][0])
+
+    def test_creating_an_ipaddress_without_namespace_or_parent(self):
+        with self.assertRaises(ValidationError) as err:
+            ip = IPAddress(address="1976:2023::1/128", status=self.status)
+            ip.validated_save()
+        self.assertEqual(err.exception.message_dict["parent"][0], "Namespace could not be determined.")
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_models.py:980]: Maybe this error should be changed to something like `"Either a parent or a namespace must be provided"`?

**Review by glennmatthews — APPROVED** [review]
A few nits about message wording, otherwise looks solid.
